lib/config/presets/index.ts

```diff
@@ -384,11 +407,48 @@ export async function resolveConfigPresets(
         val as RenovateConfig,
         baseConfig,
         ignorePresets,
-        existingPresets
+        existingPresets,
+        shallowResolve
       );
     }
   }
   logger.trace({ config: inputConfig }, 'Input config');
   logger.trace({ config }, 'Resolved config');
+  if (shallowResolve) {
+    handleExtendsArray(config, unresolvedPreserts);
+  }
   return config;
 }
+
+// handleExtendsArray:
+// This function edits extends array presets when shallowResolve is enabled
+// 1. remove resolved external presets from the extends array.
+// 2. clean duplicate presets keys (from different resolution levels).
+function handleExtendsArray(
+  config: AllConfig,
+  unresolvedPreserts: string[]
+): void {
+  if (config.extends === undefined || config.extends.length === 0) {
+    return;
+  }
+  config.extends = config.extends?.filter((e) => !shouldShallowResolve(e));
+  config.extends?.push(...unresolvedPreserts);
+  const uniqueExtends = new Set(config.extends?.values());
+  config.extends = Array.from(uniqueExtends);
+  if (config?.extends?.length === 0) {
+    // clean empty extends array
+    delete config.extends;
+  }
+}
+
+export function shouldShallowResolve(presetSource: string): boolean {
+  if (whitesourcePresetRegex.test(presetSource)) {
+    return false;
+  }
+  return (
+    presetSource.startsWith('github>') ||
+    presetSource.startsWith('gitlab>') ||
+    presetSource.startsWith('gitea>') ||
+    presetSource.startsWith('local>')
```

Comment:
`local>` is optional, so a presest which contains a `/` but not `@` of `:` is also considered local
https://github.com/renovatebot/renovate/blob/3c8d1d6076e2e776ebc366274e97198d9e69bce4/lib/config/presets/index.ts#L120-L126